### ISSUE TITLE
Fitted model doesn't use given class_weights (or scale_pos_weight) in predict/predict_proba

### ISSUE THREAD

**Issue by ranmee** [catboost/catboost#1135]
Hi,
I'm fitting a model with both categorical and numerical data (also happens for just categorical data) and I pass **class_weights**. Afterwards, when I calculate a metric, in my case: Precision, with the predictions of the predict function (or predict_proba), the result I get is for the model without using the class weights I've provided. 
Am I doing something wrong? Is this a bug? Should I set some parameter so the fitted model will use the class weights I've set?
Btw, this also happens if I use scale_post_weight. 

I have constructed an example to illustrate this, bases on the catboost pydata London 2019 notebook:
```
from catboost.datasets import amazon
from catboost import CatBoostClassifier
from catboost import Pool
from sklearn.model_selection import train_test_split

# Load the data.
(train_df, test_df) = amazon()

# Set X and y.
y = train_df.ACTION
X = train_df.drop('ACTION', axis=1)

# Declare categorical features.
cat_features = list(range(0, X.shape[1]))
print(cat_features)

# Split the data into train and test.
data = train_test_split(X, y, test_size=0.2, random_state=0)
X_train, X_validation, y_train, y_validation = data

# Create two pools for train and validation.
train_pool = Pool(data=X_train, label=y_train, cat_features=cat_features)
validation_pool = Pool(data=X_validation, label=y_validation, cat_features=cat_features)

# Prepare the model. Notive I've set class_weights (the numbers are just for demonstrating the issue).
# I've also set use_best_model to False so it won't chop off any trees according to the logloss function score.
catboost_model = CatBoostClassifier(
    iterations=300,
    learning_rate=0.01,
    custom_loss=['AUC', 'Accuracy', 'Precision', 'Recall'],
    class_weights=[0.94, 0.06],
    use_best_model=False,
    random_seed=42
)

# Fit the model.
catboost_model.fit(train_pool, eval_set=validation_pool, verbose=False, plot=True)
# If you run this you will get:
# In the "Precision:use_weights=true" we can see the test set gets 0.8087 for the final tree.
# In the "Precision:use_weights=false" we can see the test set gets 0.9851 for the final tree.
```

Then I calculate the precision score on the evaluation set:
```
# Use sklearn to calculate metrics.
from sklearn import metrics

# Run predict. Same behaviour happens for predict_proba as well.
predictions = catboost_model.predict(validation_pool)

# Calculate precision.
precision = metrics.precision_score(y_validation, predictions)
print(f"Precision: {precision}")
# prints: Precision: 0.9851281132413546

# As we can see the score is identical to the one of "Precision:use_weights=false".
# How do I predict with the weighted model??
```

I tried to find an answer for this, since I really need the results from the model using the class weights, but wasn't able to find any solution :(
Thanks in advance for you help!

**Comment by annaveronika** [comment]
Predictions do not use weights, metric calculation should use weights. You are not passing weights to metric.precision_score, so you have not weighted score.

One option is to use catboost eval_metrics method, where you can pass weights to metric calculation. Use metric_period={number of iterations} to calculate it only on the last iteration.

Hope it helps! Let us know if you have more questions.

**Comment by ranmee** [comment]
@annaveronika thanks for your reply!
I’m not I understand. I thought the point of the class_weights and scale_pos_weight parameters is to make the model pay more attention to a certain class so imbalanced dataset could be handled, but from your answer it seem to only affect the metric.
In my case, I have a highly imbalanced data for binary classification, so I want to train my model accordingly, and when a real object arrives I run predict without a weight (since I don’t know what class it belongs to).
So I guess my question is how to use catboost for highly imbalanced data? I would like to use weights so the minority class has a higher weight thus the model is more careful with its prediction.
Thanks again!

**Comment by annaveronika** [comment]
class_weights and scale_pos_weight assign a particular weight to the samples of particular class. The weights are used in the training in the following way: instead of minimizing the initial loss function, the weighted loss function is minimized. This way you have a model that pays more attention to some class then the other, for example you might make the model pay more attention to the underrepresented class in your unbalanced dataset.
But the algorithm under the class_weights is very simple, it is reweighting of the samples for the optimized loss function.

**Comment by ranmee** [comment]
@annaveronika so why when I later use predict on the same validation set I get different results (say the precision score)? The results I get represent the unweighted model rather for the trained weighted model.
My point is, if the model was trained using the weights in the loss function, why are the results different when using predict?

**Comment by annaveronika** [comment]
When you are calculating metric after training you are calculating a metric without weights. When you are calculating metric during training, we calculate both metrics - with weights and without weights. If you are interested in the not weighted one, you need to look on the one with parameter use_weights=False.

**Comment by Younes7777** [comment]
Hello,

I have almost the same problem, 
I am struggling with an Imbalanced data (target is binary):

print(class_weights)
{0: 1.0, 1: 5.85}

params = {'loss_function':'Logloss', # objective function
          'eval_metric':'AUC', # metric
          'iterations':2000,
          'cat_features': cat_features,
          'early_stopping_rounds': 120,
           'custom_loss': ['AUC','Accuracy' ,'Precision', 'Recall'],
          #'learning_rate' : 0.15, 
          'scale_pos_weight': class_weights[1],
          'verbose' : 50,
          'use_best_model': True
         }

cbc_1 = CatBoostClassifier(**params)
cbc_1.fit(train_pool,eval_set=(test_pool),use_best_model=True,plot=True)

i want to get the confusion_matrix:
i use:
get_confusion_matrix(cbc_1, Pool(X_test, y_test, cat_features=cat_features))

but I get a matrix without weights, as if I trained the model without scale_pos_weight. There are weight parameters in get_confusion_matrix, but when I enter the weight (scale_pos_weight ), I get an error message. 
does anyone know how i can get the confusion matrix with the weights? that would be great.
Thanks in advance
Younes

**Comment by jboverio** [comment]
The answers are not clear...

**Comment by Theghios** [comment]
The question has not been answered and the problem persists.
It is not clear whether the model actually weights more the under represented class and how the metrics are calculated.
Quoting @annaveronika 

> When you are calculating metric after training you are calculating a metric without weights. When you are calculating metric during training, we calculate both metrics - with weights and without weights. If you are interested in the not weighted one, you need to look on the one with parameter use_weights=False.

Does that mean that only the metrics are weighted but not the actual samples during training? 
Thanks for your answers, I believe this is quite an important problem for catboost to solve.
